tests/storage_test.go

```diff
@@ -245,6 +245,133 @@ var _ = Describe("Storage", func() {
 			})
 
 		})
+		Context("with an Fedora shared NFS PVC, cloud init and service account", func() {
+			var pvName string
+			var hugepageSize = "2Mi"
+			var vmi *v1.VirtualMachineInstance
+			var nodeWithHugepages *k8sv1.Node
+			var hugepageType k8sv1.ResourceName
+			var originalPagesCount resource.Quantity
+			var shouldUpdatePagesCount = false
+			BeforeEach(func() {
+				hugepageType = k8sv1.ResourceName(k8sv1.ResourceHugePagesPrefix + hugepageSize)
+
+				By("Looking for a node with hugepages")
+				nodeWithHugepages = tests.GetWorkerNodeWithHugepages(virtClient, hugepageType)
+				if nodeWithHugepages == nil {
+					Skip(fmt.Sprintf("No node with hugepages %s capacity", hugepageType))
+				}
+
+				originalPagesCount, _ = nodeWithHugepages.Status.Capacity[hugepageType]
+				desiredAmountOfHugePages := resource.MustParse("512Mi")
+				// there is no need to update the number of host hugepages if there is a satisfying amount
+				if originalPagesCount.Cmp(desiredAmountOfHugePages) < 0 {
+					shouldUpdatePagesCount = true
+				}
+
+				if shouldUpdatePagesCount {
+					By("Updating node capacity to 256 pages of 2M")
+					for i := 0; i < 2; i++ {
+						tests.UpdateNodeHugepages(hugepageType, 512, nodeWithHugepages.Name)
+					}
+				}
+
+				pvName = "test-nfs" + rand.String(48)
+				// Prepare a NFS backed PV
+				By("Starting an NFS POD")
+				os := string(tests.ContainerDiskFedora)
+				nfsIP := tests.CreateNFSTargetPOD(os)
+				// create a new PV and PVC (PVs can't be reused)
+				By("create a new NFS PV and PVC")
+				tests.CreateNFSPvAndPvc(pvName, "300Mi", nfsIP, os)
+				tests.EnableFeatureGate(virtconfig.VirtIOFSGate)
+			})
+
+			AfterEach(func() {
+				tests.DisableFeatureGate(virtconfig.VirtIOFSGate)
+				By("Deleting the VMI")
+				Expect(virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})).To(Succeed())
+
+				By("Waiting for VMI to disappear")
+				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+				// PVs can't be reused
+				tests.DeletePvAndPvc(pvName)
+
+				By("Deleting NFS pod")
```

Comment:
Here's a thought that might help simplify the tests a bit, what would you think about using a pvc dynamically created by the default storageclass rather than using an NFS pod?

For example, If you look at some of the datavolume tests, you could create a VM with a DataVolumeTemplate (which automatically creates the PV/PVC for us) the assign that PVC to the filesystem in the VMI template spec.

For validation, you could have cloud-init mount virtiofs and also write some file to the directory after the mount occurs. Then you could podExec into virt-launcher to verify the file is visible outside of the guest.


That would remove the necessity for creating the nfs pod and nfs related pv/pvcs directly.

Reply:
I've been looking at the datavolume tests and honestly didn't make anything out of it... :-(

I think NFS based test is now much easier. It also reflects a real-life example of how I think users may use it.

Reply:
@davidvossel I've switched to data volume ;)